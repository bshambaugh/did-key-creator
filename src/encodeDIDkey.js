import * as u8a from 'uint8arrays'
import multicodec from 'multicodec'
import  multibase from'multibase'

export function encodeDIDfromHexString(publicKeyHex: string, multicodecName: string) {
    const publicKey = u8a.fromString(publicKeyHex,'base16');
    const didKey = encodeDIDfromBytes(publicKey,multicodecName);
    return didKey;
}

export function encodeDIDfromBytes(publicKey: Uint8Array, multicodecName: string) {
          const publicKeywPrefix = multicodec.addPrefix(multicodecName,publicKey)
          const buf = multibase.encode('base58btc',publicKeywPrefix)
          const bufAsString = u8a.toString(buf)
	  return `did:key:${bufAsString}`	  
}
