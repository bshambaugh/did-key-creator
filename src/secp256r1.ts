import * as u8a from 'uint8arrays'
import multicodec from 'multicodec'
import  multibase from'multibase'

export function encodeDIDfromHexString(publicKeyHex: string) {
    const publicKey = u8a.fromString(publicKeyHex,'base16');
    const didKey = encodeDIDfromBytes(publicKey)
    return didKey;
}

export function encodeDIDfromBytes(publicKey: Uint8Array) {
          const publicKeywPrefix = multicodec.addPrefix('p256-pub',publicKey)
          const buf = multibase.encode('base58btc',publicKeywPrefix)
          const bufAsString = u8a.toString(buf)
	  return `did:key:${bufAsString}`	  
}
