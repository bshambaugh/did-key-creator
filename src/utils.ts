import * as u8a from 'uint8arrays'
import multicodec from 'multicodec'
import  multibase from'multibase'

// compress a public key with points x,y expressed as UintArrays
// source: https://stackoverflow.com/questions/17171542/algorithm-for-elliptic-curve-point-compression
export function ECPointCompress( x: Uint8Array, y: Uint8Array )
{
    const out = new Uint8Array( x.length + 1 );

    out[0] = 2 + ( y[ y.length-1 ] & 1 );
    out.set( x, 1 );

    return out;
}

// create a compressed public key in hex from a raw public key in hex
export function compressedKeyInHexfromRaw(publicKeyHex: string) {
  const xHex = publicKeyHex.slice(0,publicKeyHex.length/2);
  const yHex = publicKeyHex.slice(publicKeyHex.length/2,publicKeyHex.length);

  const xOctet = u8a.fromString(xHex,'base16')
  const yOctet = u8a.fromString(yHex,'base16');

  const compressedPoint = ECPointCompress( xOctet , yOctet );
  const compressedPointHex = u8a.toString(compressedPoint,'base16');
  return compressedPointHex;
}

// create an uncompressed public key from a raw public key in hex
export function uncompressedKeyInHexfromRaw(publicKeyHex: string) {
   return '04'+publicKeyHex;
}

// create a raw public key from a uncompressed public key in hex
export function rawKeyInHexfromUncompressed(publicKeyHex: string) {
   return publicKeyHex.slice(2);
}

export function pubKeyHexToUint8Array(publicKeyHex: string) {
  if(publicKeyHex == null) {
   throw new TypeError('input cannot be null or undefined.');
  }
    if(publicKeyHex.length % 2 == 0) {
          return u8a.fromString(publicKeyHex,'base16');
      } else {
          return u8a.fromString(('0'+publicKeyHex),'base16');
    }
}

export function didKeyIDtoPubKeyHex(didKeyID : string) {
   const buf = multibase.decode(didKeyID);
   const bufwoPrefix = multicodec.rmPrefix(buf);
   return u8a.toString(bufwoPrefix,'base16')
}

export function didKeyURLtoPubKeyHex(didKeyURL : string) {
    const didKeyID = didKeyURL.split(':')[2];
    return didKeyIDtoPubKeyHex(didKeyID);
}
