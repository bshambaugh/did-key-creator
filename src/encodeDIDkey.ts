import * as u8a from 'uint8arrays'
import multicodec from 'multicodec'
import  multibase from'multibase'
import { CodecName } from 'multicodec/src/generated-types';


export function encodeDIDfromHexString(multicodecName: CodecName,publicKeyHex: string) {
    const publicKey = u8a.fromString(publicKeyHex,'base16');
    const didKey = encodeDIDfromBytes(multicodecName,publicKey);
    return didKey;
}


export function encodeDIDfromBytes(multicodecName: CodecName, publicKey: Uint8Array) {
          const publicKeywPrefix = multicodec.addPrefix(multicodecName,publicKey)
          const buf = multibase.encode('base58btc',publicKeywPrefix)
          const bufAsString = u8a.toString(buf)
	  return `did:key:${bufAsString}`	  
}
