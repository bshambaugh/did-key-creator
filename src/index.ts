import { ECPointCompress,
  compressedKeyInHexfromRaw, 
  uncompressedKeyInHexfromRaw, 
  rawKeyInHexfromUncompressed, 
  pubKeyHexToUint8Array, 
  didKeyIDtoPubKeyHex, 
  didKeyURLtoPubKeyHex } from './did_key_utils.js'

import { encodeDIDfromHexString, encodeDIDfromBytes } from './encodeDIDkey.js'

export { ECPointCompress,
         compressedKeyInHexfromRaw, 
         uncompressedKeyInHexfromRaw, 
         rawKeyInHexfromUncompressed, 
         pubKeyHexToUint8Array, 
         didKeyIDtoPubKeyHex, 
         didKeyURLtoPubKeyHex,
         encodeDIDfromHexString, 
         encodeDIDfromBytes
       } 
