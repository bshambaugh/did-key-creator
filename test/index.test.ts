import u8a from 'uint8arrays'

import { encodeDID } from '../src'

describe('key-did-provider-secp256r1', () => {
  test('encodeDID', () => {
    const key = u8a.fromString(
      'f9c36f8964623378bdc068d4bce07ed17c8fa486f9ac0c2613ca3c8c306d7bb61cd36717b8ac5e4fea8ad23dc8d0783c2318ee4ad7a80db6e0026ad0b072a24f',
      'base16'
    )
    expect(encodeDID(key)).toMatchSnapshot()
  })
})
