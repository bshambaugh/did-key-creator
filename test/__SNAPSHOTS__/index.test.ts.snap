// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`key-did-provider-secp256r1 encodeDID 1`] = `"did:key:zruuPojWkzGPb8sVc42f2YxcTXKUTpAUbdrzVovaTBmGGNyK6cGFaA4Kp7SSLKecrxYz8Sc9d77Rss7rayYt1oFCaNJ"`;
